src/diffusers/models/attention_processor.py

```diff
@@ -2220,25 +2225,37 @@ def __init__(self, hidden_size, cross_attention_dim=None, num_tokens=4, scale=1.
 
         self.hidden_size = hidden_size
         self.cross_attention_dim = cross_attention_dim
+
+        if not isinstance(num_tokens, list):
+            raise ValueError("`num_tokens` should be a list of integers.")
         self.num_tokens = num_tokens
+
+        if not isinstance(scale, list):
+            scale = [scale] * len(num_tokens)
+        if len(scale) != len(num_tokens):
+            raise ValueError("`scale` should be a list of integers with the same length as `num_tokens`.")
         self.scale = scale
 
-        self.to_k_ip = nn.Linear(cross_attention_dim or hidden_size, hidden_size, bias=False)
-        self.to_v_ip = nn.Linear(cross_attention_dim or hidden_size, hidden_size, bias=False)
+        self.to_k_ip = nn.ModuleList(
+            [nn.Linear(cross_attention_dim, hidden_size, bias=False) for _ in range(len(num_tokens))]
+        )
+        self.to_v_ip = nn.ModuleList(
+            [nn.Linear(cross_attention_dim, hidden_size, bias=False) for _ in range(len(num_tokens))]
+        )
 
     def __call__(
         self,
         attn,
         hidden_states,
-        encoder_hidden_states=None,
+        encoder_hidden_states,
         attention_mask=None,
         temb=None,
-        scale=1.0,
+        **kwargs,
```

Comment:
Let's not add a kwargs here unnecessarily

Reply:
IMO it's not unnecessary and all processors should use this layout as it is much more flexible and less error prone if cross_attention_kwargs are actually used. At the moment it's not possible to use a processor with custom signature without awkward cross_attention_kwargs dict restructuring in attention.py, a core file, as cross_attention_kwargs has to match all receiving function signatures.

Reply:
I agree that we don't need `**kwargs` here, but I'm slightly in favor of adding it - I think it's more readable and makes mix-and-match attention processors easier here. 

For example, the `scale` argument is not used in the IP-adapter attention calculation at all. However, we have to add it in the signature; otherwise, it won't work because the IP-adapter attention processor is used together with the default attention processors, which takes a `scale` argument. 

Likewise, if we want to add a new argument to ip-adapter attention processor, we would have to add that argument to the default attention_processor's signature too (e.g. if we are going to support ip-adapter attention mask, we will have to add a new `ip-attention-mask` argument) - I think it is more confusing than simply appending a `**kwargs` to the signature.

Reply:
I agree with Yiyi here.

Reply:
If you feel strongly here @yiyixuxu go for it! Still don't think we should add it though because adding `kwargs` makes it much harder for people to spot errors. E.g. if a function has a **kwarg input, it swallows everything.

E.g. if you have:
```py
def forward(input_values, attention_scale=1.0, **kwargs):
      ...
```

and someone write `forward(input_values, atention_scale=5.0)` # note the typo for attention_scale here.

Then this code is executed nevertheless with `attention_scale` set to 1.0 though because `atention_scale` is swallod by kwargs.

Most programming languages don't allow something like `**kwargs` for good reason in my opinion. 
Also it sets a bad precedent and incentivizes people to add kwargs everywhere.
It also significantly hurts readability. Having a kwargs argument means that you can never be sure what can all be passed to the function making it harder to understand the code. 

But again, I do see the upsides you mention above. Another possibility that we could start to explore more is to add `validation` function decorators that are used quite heavily in `huggingface_hub`: https://github.com/huggingface/huggingface_hub/blob/c528f78fa7404869f32a2b04eac0f9964a5c6f9e/src/huggingface_hub/utils/_validators.py#L46 we could use such a validator to remove an argument like `scale`

Reply:
Using validation decorators would be very nice to have. I am happy to take a look at it next week.

Reply:
thanks for the explanation! make sense! I will not add `**kwargs` for now:)

Reply:
I posted a comment [here](https://github.com/huggingface/diffusers/issues/6814#issuecomment-1935547992) to complement on @patrickvonplaten's comment above. IMO using a decorator to avoid **kwargs is legit for some use cases while more debatable in other situations. (just to keep it in mind)